<a href="https://colab.research.google.com/github/duckslayer101/Basic-DS-Theories/blob/main/Array.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# ***Array***



*   Array is matrix



*   Homogenous (can only store one data type at a time)
    (There is exception also)
*   Has fixed length


*   More efficient than list
*   Have to import to use this library











In [135]:
import numpy as np

# **Linear Array**

In [179]:
# initialization

a = np.array([0]*4)
print(a)

[0 0 0 0]


In [136]:
a = np.array([1, 2, 3])
print(a)

# array's values are changable
a[1] = 4
print(a)

[1 2 3]
[1 4 3]


In [137]:
# empty value/ null value/ dummy value will be in right side.

arr = np.array([1, 2, 3, 0])

# Here 0 is a dummy value.

In [138]:
"""In order to put string into another array we have to use None cause
    if we do [''] this will take the first character only. That's why
    have to do [None] in order to take string.
"""

a = np.array(['abc', 'def'])
b = np.array([None]*2)
for i in range(len(a)):
    b[i] = a[i]
print(b)

['abc' 'def']


In [139]:
# Iteration

for i in a:
  print(i)

for i in range(len(a)):
  print(a[i])

abc
def
abc
def


In [140]:
# Specail initilization where the values of array are treated like object

a1 = np.array([None]*5)
print(a1)

a1[2] = 3
print(a1)

# Here because of None the array is actually working as a non-homogenous since it's treating the none values as object.

[None None None None None]
[None None 3 None None]


In [141]:
# array is matrix. In this addition scenario it adds like matrix addition. It doesn't add like in list.

b = np.array([1, 2, 3])
c = np.array([4, 5, 6])

n = b+c
print(n)

[5 7 9]


In [142]:
# Resize. In order to resize an array, we need to create a new array with the new length and copy values from old array to the new array.

arr = np.array([1, 2, 3, 4])
new_arr = np.zeros(8, dtype=int)

for i in range(len(arr)):
  new_arr[i] = arr[i]
print(new_arr)

[1 2 3 4 0 0 0 0]


In [143]:
# another way of initializing

arr = np.zeros(5)
print(arr)

# Here the default data type is float. It can be changed by using dtype='....'

arr = np.zeros(5, dtype=str)
print(arr)

[0. 0. 0. 0. 0.]
['' '' '' '' '']


In [144]:
a = np.array([1, 2, 3, 4, 5])
print(a[-1::-1])

# Reverse iteration

for i in range(len(a)-1,-1,-1):
    print(a[i])

[5 4 3 2 1]
5
4
3
2
1


In [145]:
# copy in array (pass by reference)

a = np.array([1, 2, 3])
b = a
print(b)
b[1] = 4
print(a)

[1 2 3]
[1 4 3]


In [146]:
# deep copy (pass by value)

a = np.array([1, 2, 3])
b = a.copy()
print(b)
b[1] = 5
print(a)
print(b)

[1 2 3]
[1 2 3]
[1 5 3]


In [147]:
# left shift        [2 3 0]

a = np.array([1, 2, 3])
def left_shift(arr):
    for i in range(len(arr)-1):
        arr[i] = arr[i+1]
    arr[-1] = 0
    return arr

print(left_shift(a))

[2 3 0]


In [148]:
# right shift       [0 1 2]

a = np.array([1, 2, 3])
def right_shift(arr):
    for i in range(len(arr)-1, 0, -1):
        arr[i] = arr[i-1]
    arr[0] = 0
    return arr

print(right_shift(a))

[0 1 2]


In [149]:
# left rotate       [2 3 1]

a = np.array([1, 2, 3])
def left_rotate(arr):
    temp = arr[0]
    for i in range(len(arr)-1):
        arr[i] = arr[i+1]
    arr[-1] = temp
    return arr

print(left_rotate(a))

[2 3 1]


In [150]:
# right rotate      [3 1 2]

a = np.array([1, 2, 3])
def right_rotate(a):
    temp = a[-1]
    for i in range(len(a)-1, 0, -1):
        a[i] = a[i-1]
    a[0] = temp
    return a

print(right_rotate(a))

[3 1 2]


In [151]:
# Reverse (out of place)

a = np.array([1, 2, 3, 4, 5])
b = np.array([0]*len(a))
for i in range(len(b)):
    b[i] = a[len(a)-i-1]
print(b)

[5 4 3 2 1]


In [152]:
# Reverse (in place)

a = np.array([1, 2, 3, 4, 5, 6, 7])
def reverse_inplace(a1):
    for i in range(len(a1)//2):
        a1[i], a1[len(a1)-i-1] = a1[len(a1)-i-1], a1[i]
    return a1

print(reverse_inplace(a))
print(a)

[7 6 5 4 3 2 1]
[7 6 5 4 3 2 1]


In [153]:
a = 1
b = 2
a, b = b, a
print(a, b)

2 1


In [154]:
# Size of array. It's the number of non-dummy values in an array. It's used in insertion and in delation

a = np.zeros(5, dtype=int)
print(a)
a[0], a[1], a[2] = 1, 2, 3
def non_dummies(arr, size):
    for i in range(size):
        print(arr[i])

non_dummies(a, 3)
print(a)

[0 0 0 0 0]
1
2
3
[1 2 3 0 0]


In [155]:
# insert (at the end)

a = np.array([1, 2, 3, 4, 5, 6])
def resize_array(arr, new_length):
    a = np.zeros(new_length, dtype= int)
    for i in range(len(arr)):
        a[i] = arr[i]
    return a

def insert_athe_end(arr, size, value):
    if size >= len(arr):
        arr = resize_array(arr, len(arr)*2)
    arr[size] = value
    return arr


print(insert_athe_end(a, 6, 7))

[1 2 3 4 5 6 7 0 0 0 0 0]


In [156]:
a = np.array([1, 2, 3, 4])
def insertend(a, size, elem):
    if size >= len(a):
        a = resize_array(a, len(a)*2)
    a[size] = elem
    return a

print(insertend(a, 4, 5))

[1 2 3 4 5 0 0 0]


In [157]:
# insert anywhere but end

a =  np.array([1, 2, 3, 4, 5])
def insert_anywhere_butend(a, size, index, elem):
    if index<0 or index>=len(a):
        return 'Inserting is not possible'
    if size >= len(a):
        a = resize_array(a, len(a)*2)
        print(f'Resized array: {a}')
    for i in range(len(a)-1, index, -1):
        a[i] = a[i-1]
    a[index] = elem
    return a

print(insert_anywhere_butend(a, size = 5, index = 2, elem = 10))

Resized array: [1 2 3 4 5 0 0 0 0 0]
[ 1  2 10  3  4  5  0  0  0  0]


In [158]:
a = np.array([1, 2, 3, 5, 6, 0])
def insert_inbetween(arr, size, index, elem):
    if index<0 or index>=len(arr):
        return 'Not possible'
    if size >= len(arr):
        arr = resize_array(arr, len(arr)*2)
    for i in range(len(arr)-1, index, -1):
        arr[i] = arr[i-1]
    arr[index] = elem
    return arr

print(insert_inbetween(a, 5, 3, 4))

[1 2 3 4 5 6]


In [159]:
# Remove last element

arr = np.array([1, 2, 3, 4])
def rmv_last_elem(arr, size):
    if size == 0:
        return 'Remove isn\'t possible'
    arr[size-1] = 0
    return arr

print(rmv_last_elem(arr, 4))

[1 2 3 0]


In [160]:
arr = np.array([1,2,3,4,5,6,7,0,0,0])
def rmvlast(arr, size):
    if size == 0:
        return 'Not possible'
    arr[size-1] = 0
    return arr

print(rmvlast(arr, 7))

[1 2 3 4 5 6 0 0 0 0]


In [161]:
# Remove any other but last elem

def elem_inbetween(arr, size, index):
    if index<0 or index>=len(arr):
        return 'Not possible'
    arr[index] = 0
    for i in range(index, size-1):
        arr[i] = arr[i+1]
    arr[size-1] = 0
    return arr


a = np.array([1, 2, 3, 4, 5])
print(elem_inbetween(a, 5, 1))

[1 3 4 5 0]


In [162]:
def elem_remv(arr, index, size):
    if index<0 or index>=len(arr):
        return 'Not possible'
    arr[index] = 0
    for i in range(index, len(arr)-1):
        arr[i] = arr[i+1]
    arr[size-1] = 0
    return arr

arr = np.array([1, 2, 3, 4, 5])
print(elem_remv(arr, index = 2, size = 5))

[1 2 4 5 0]


In [163]:
# Sorting array
# small to big

def small_to_big(arr):
    for i in range(len(arr)-1):         # ***This (len-1) is so important***
        min = arr[i]
        for j in range(i+1, len(arr)):
            if arr[j]<min:
                min = arr[j]
                indx = j
        arr[i], arr[indx] = min, arr[i]
    return arr
arr = np.array([5, 3, 6, 8, 2, 0, -1, -2])
print(small_to_big(arr))

[-2 -1  0  2  3  5  6  8]


In [183]:
def small_to_bg(arr):
    for i in range(len(arr)-1):
        min = arr[i]
        for j in range(i+1, len(arr)):
            if arr[j]<min:
                min = arr[j]
                indx = j
        arr[i], arr[indx] = min, arr[i]
        print(arr)
    return arr

a = np.array([1, 0, 4, 2, -1, -5, 7])
print(small_to_bg(a))

[-5  0  4  2 -1  1  7]
[-5 -1  4  2  0  1  7]
[-5 -1  0  2  4  1  7]
[-5 -1  0  1  4  2  7]
[-5 -1  0  1  2  4  7]
[-5 -1  0  1  2  4  7]
[-5 -1  0  1  2  4  7]


# **Multidimensional Array**

*   2D Array
*   3D Array


*   There could be more dimensional array too


*   Dimensions have to be fixed









In [165]:
# Initialization

arr = np.array([[1, 2, 3], [4, 5, 6]])
print(arr)

[[1 2 3]
 [4 5 6]]


In [166]:
# Another way of initialization

m = np.zeros((3,2), dtype=int)      # same as linear array
print(m)

[[0 0]
 [0 0]
 [0 0]]


In [167]:
# row = length of outer array
print(f'Row: {len(arr)}')

# column = length of inner array
print(f'Column: {len(arr[0])}')


Row: 2
Column: 3


In [168]:
# another way of finding row and column

row, column = arr.shape     # This gives a tuple
print(row, column)          # The 1st one is row & the 2nd one is column

2 3


In [169]:
# Iteration         # row print also

arr = np.array([[1, 2, 3], [4, 5, 6]])
for i in range(len(arr)):
    for j in range(len(arr[0])):
        print(arr[i][j])

1
2
3
4
5
6


In [170]:
# column print

m = np.array([[1, 2, 3], [4, 5, 6]])
print(m)

def column_print(m):
    r, c = m.shape
    for i in range(c):
        for j in range(r):
            print(m[j][i])

column_print(m)

[[1 2 3]
 [4 5 6]]
1
4
2
5
3
6


In [171]:

m = np.array([[1, 2, 3], [4, 5, 6]])
for i in range(len(m[0])):
    for j in range(len(m)):
        print(m[j][i])

1
4
2
5
3
6


In [172]:
# Add 2 matrix

def add_matrix(m, n):
    if m.shape != n.shape:
        return 'Not possible'
    for i in range(len(m)):
        for j in range(len(m[0])):
            m[i][j] += n[i][j]
    return m

a = np.array([[1,2,3], [4,5,6]])
b = np.array([[1,1,1], [1,1,1]])
print(add_matrix(a, b))
print(a)

[[2 3 4]
 [5 6 7]]
[[2 3 4]
 [5 6 7]]


In [173]:
# Short way of adding. It's one of the advantages of numpy

a = np.array([[1,2,3], [4,5,6]])
b = np.array([[1,1,1], [1,1,1]])
print(a+b)

[[2 3 4]
 [5 6 7]]


In [174]:
# You can change value by doing this short way too.

a = np.array([[1,2,3], [4,5,6]])
print(a*10)

[[10 20 30]
 [40 50 60]]


In [175]:
# Multiplication of two 2D Array
                                                           # This one is hard:)
""" The result of these arrays is:
    [19 22]
    [43 50] """

# The length of the result array will be (row of 1st array) x (column of 2nd array)

""" The condition is: (column of 1st array) == (row of 2nd array).
    otherwise the multiplication won't be possible. """


def multiplication(m, n):
    rm, cm = m.shape
    rn, cn = n.shape
    if cm != rn:
        return 'error'
    new_arr = np.zeros((rm, cn), int)
    for i in range(rm):
        for j in range(cn):
            for k in range(cm):             #this one is the index of new_arr
                new_arr[i][j] += m[i][k]*n[k][j]
    return new_arr

m = np.array([[1,2], [3,4]])
n = np.array([[5,6], [7,8]])
print(multiplication(m, n))
print()
m1 = np.array([[3,1,2]])
n1 = np.array([[2,1], [1,2], [4,7]])
print(multiplication(m1, n1))

[[19 22]
 [43 50]]

[[15 19]]


In [176]:
# find sum of selected row and column
# this one is (5)+(3+1)+(1+2+1)

def find_sum(m):
    sum = 0
    for i in range(len(m)):
        for j in range(i+1):
            sum += m[i][j]
    return sum

m = np.array([[5,2,1], [3,1,2], [1,2,1]])
print(find_sum(m))

13


In [198]:
# sum of every row

def sum_row_wise(m):
    r, c = m.shape
    result = np.zeros((r, 1), int)
    for i in range(r):
        for j in range(c):
            result[i][0] += m[i][j]
    return result

m = np.array([[4,3,8], [2,5,1], [7,-1,9]])
print(sum_row_wise(m))

[[15]
 [ 8]
 [15]]


In [200]:
# Swap the column of a matrix

def swap_column(m):
    r, c = m.shape
    for i in range(r-1):
        for j in range(c//2):
            m[i][j], m[i][c-j-1] = m[i][c-j-1], m[i][j]
    return m

m = np.array([[4,3,8], [2,5,-1], [-6,16,9], [4,13,11]])
print(swap_column(m))

[[ 8  3  4]
 [-1  5  2]
 [ 9 16 -6]
 [ 4 13 11]]


In [202]:
# sum of the elements of (primary diagonal) in a (square matrix)
# meaning here it is 4+5+9

def sum(arr):
    sum = 0
    r, c = arr.shape
    if r!=c:
        return 'error'
    for i in range(r):
        sum += arr[i][i]
    return sum

m = np.array([[4,3,8], [2,5,1], [7,6,9]])
print(m)
print(sum(m))

[[4 3 8]
 [2 5 1]
 [7 6 9]]
18


In [203]:
# sum of the elements of (secondary diagonal) in a (square matrix)
# meaning here it is 8+5+7

def summ(m):
    r, c = m.shape
    if r!=c:
        return 'error'
    sum = 0
    for i in range(r):
        sum += m[i][r-i-1]
    return sum

m = np.array([[4,3,8], [2,5,1], [7,6,9]])
print(m)
print(summ(m))

[[4 3 8]
 [2 5 1]
 [7 6 9]]
20


# ***Short Questions***

1. **How to find dimension index of linear index**

    Que : Suppose a linear array of length 128 actually stores a 3D array of dimensions 4x4x8.
    What are the multidimensional indexes of the element stored at location 111?

    Ans :

    x = 111 // (4x8) = 3

    111 % (4x8) = 15

    y = 15 // 8 = 1

    15 % 8 = 7

    z = 7 // 1 = 7

    So the answer is [3][1][7]
    
    ---
    vice versa

    Linear index = 3x4x8 + 1x8 + 7 = 111
    
    So the answer is 111